### PR TITLE
SLING-7534 generate sha-512 hashes by default and also attach to build

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

File: sling-parent/pom.xml [hunk 1]
```diff
@@ -108,6 +108,165 @@
...
+                    <executions>
+                        <execution>
+                            <id>generate-depends-file</id>
+                            <goals>
+                                <goal>generate-depends-file</goal>
+                            </goals>
+                        </execution>
+                    </executions>
+                </plugin>
+                <!-- generate sha512 checksums for all attached artifacts (not supported by ASF parent pom yet, https://issues.apache.org/jira/browse/MPOM-244) -->
```
Comment (kwin) [inline sling-parent/pom.xml:248]: pluginMgmt only reordered (place before plugins). Only this plugin is actually new/touched!

Comment (rombert) [reply sling-parent/pom.xml:248]: Would it make sense then to split in this change two commits - one with reordering, one with applying the actual changes? Makes it easier to track in time.

Comment (kwin) [reply sling-parent/pom.xml:248]: done, reordering in https://github.com/apache/sling-parent/pull/10/commits/571fba919627a7751701d434ea147927841652d3, actual changes in https://github.com/apache/sling-parent/pull/10/commits/303b3e5e96592d1871a8ee44efeb3ed49e436290

**Comment by kwin** [comment]
With the most recent m-deploy-plugin (3.0.0-M1) MD5/SHA1 calculation has been moved from m-install-plugin (https://issues.apache.org/jira/browse/MINSTALL-143) to m-deploy-plugin (https://issues.apache.org/jira/browse/MDEPLOY-231). Unfortunately the old parameter {{createChecksum}} is no longer available in the new m-deploy-plugin. I currently don't know how to prevent the m-deploy-plugin from generating hashes...

**Comment by kwin** [comment]
@rombert Please try again with https://github.com/apache/sling-parent/pull/10/commits/89ad8770457f7faf88f01107ccc5d9ffc34179c2

**Comment by rombert** [comment]
@kwin  - looks the same to me:
```
$ tree target/repo/
target/repo/
└── org
    └── apache
        └── sling
            └── org.apache.sling.engine
                ├── 2.7.3-SNAPSHOT
                │   ├── maven-metadata.xml
                │   ├── maven-metadata.xml.md5
                │   ├── maven-metadata.xml.sha1
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.jar
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.jar.md5
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.jar.sha1
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.jar.sha512
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.jar.sha512.md5
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.jar.sha512.sha1
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.pom
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.pom.md5
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1.pom.sha1
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1-sources.jar
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1-sources.jar.md5
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1-sources.jar.sha1
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1-sources.java-source.sha512
                │   ├── org.apache.sling.engine-2.7.3-20200608.131016-1-sources.java-source.sha512.md5
                │   └── org.apache.sling.engine-2.7.3-20200608.131016-1-sources.java-source.sha512.sha1
                ├── maven-metadata.xml
                ├── maven-metadata.xml.md5
                └── maven-metadata.xml.sha1
```

I ran a `mvn clean install` in the parent project, followed by a `mvn clean deploy -DaltDeploymentRepository=snapshot-repo::default::file:target/repo` in the engine project.

**Comment by kwin** [comment]
Ok, could reproduce. Older versions of the maven-deploy-plugin (prior to 3.0.0-M1)still use https://github.com/apache/maven/blob/a2b800de32cdb9adc1e64a43a0fc32e3ba878103/maven-compat/src/main/java/org/apache/maven/repository/legacy/DefaultWagonManager.java#L634 which generates those checksums by default. So it seems that there is no way around those, for the moment.

**Comment by kwin** [comment]
So to summarize the current status

> the -sources checksum has incorrect file name

Should be fixed by https://github.com/nicoulaj/checksum-maven-plugin/pull/87

> we checksum the sha1/md5 checksums
Tracked in https://issues.apache.org/jira/browse/MDEPLOY-271

> we don't checksum the pom
I guess we don't need non standard checksums here

> we don't checksum the metadata
IMHO Nexus will do that for metadata (according to https://issues.sonatype.org/browse/NEXUS-21802?focusedCommentId=939375&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-939375). For local repos this is not done which is fine. Also MD5/SHA1 would be enough IMHO for those!

**Comment by rombert** [comment]
> we don't checksum the pom
>> I guess we don't need non standard checksums here

Not sure what you mean here. Other artifacts get sha256 checksums, and the pom just md5 and sha1. Should it not also get a sha256 checksum?

**Comment by kwin** [comment]
For pom.xml it is not necessary. ASF only requires SHA-512 on `source-release.*` (built during release). IMHO we should additionally add those safer checksums to all binary artifacts. For pom.xml and metadata I don't care (i.e. IMHO we should stay with the default MD5/SHA1)

**Comment by kwin** [comment]
> I was curious thought, do we not see a fix "upstream", e.g. in the maven install/deploy plugins or in Nexus?
 
Although Maven Resolver added support for SHA256/512 in https://issues.apache.org/jira/browse/MRESOLVER-56 it is disabled by default for performance reasons (https://issues.apache.org/jira/browse/MRESOLVER-138). Generating the checksums with Maven Resolver requires setting system properties (https://maven.apache.org/resolver/configuration.html). Also only the upcoming Maven 4.x will include those updated Maven resolver versions.

Therefore I would recommend to stick with the approach to only generate SHA512 for binary artifacts with the https://github.com/nicoulaj/checksum-maven-plugin/ and ignore the MD5/SHA1 generated for those. Maven will internally not check against those checksums (except if explicitly configured), so we just need those primarily on the Staging repo to be able to have all artifacts for an ASF in it (without being forced to generate/transfer other artifacts when pushing to ASF dist)

**Comment by kwin** [comment]
@rombert Can you check again? We finally have a release of checksum-maven-plugin which supports the classifier used be ASF source releases.

**Comment by rombert** [comment]
@kwin - I tried this again but not sure if I'm doing this right. I rebuilt the parent project with the changes from this PR and patched the Sling Engine bundle again. 

```
$ mvn clean deploy -Papache-release -DaltDeploymentRepository=snapshot-repo::default::file:target/repo
(snip)
$ tree target/repo/
target/repo/
└── org
    └── apache
        └── sling
            └── org.apache.sling.engine
                ├── 2.7.5-SNAPSHOT
                │   ├── maven-metadata.xml
                │   ├── maven-metadata.xml.md5
                │   ├── maven-metadata.xml.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.jar
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.jar.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.jar.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.jar.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-javadoc.jar
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-javadoc.jar.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-javadoc.jar.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-javadoc.jar.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.pom
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.pom.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.pom.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.pom.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.sha512
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.sha512.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1.sha512.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-source-release.zip
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-source-release.zip.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-source-release.zip.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-source-release.zip.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-sources.jar
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-sources.jar.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.084939-1-sources.jar.md5
                │   └── org.apache.sling.engine-2.7.5-20210505.084939-1-sources.jar.sha1
                ├── maven-metadata.xml
                ├── maven-metadata.xml.md5
                └── maven-metadata.xml.sha1

5 directories, 29 files
```

The source-release does not seem to be signed, and actually he have a single `sha512` artifact without a base name. Looking at the checksums of all the files, it seems to be the checksum for the javadoc jar.

**Comment by kwin** [comment]
@rombert Unfortunately the old version of the checksum-maven-plugin was used with profile `apache-release`. This is fixed in https://github.com/apache/sling-parent/pull/10/commits/4761b82de7546704dac256de4e18be70103bfa6. Now you should see a proper SHA-512 for the source-release (and the according .asc because that has the same classifier).

**Review by rombert — APPROVED** [review]
LGTM now, thanks @kwin ! It was a bit confusing to see it work without the apache-release profile, now I understand why.

Verified that the proper signature was generated:

```
$ tree target/repo/
target/repo/
└── org
    └── apache
        └── sling
            └── org.apache.sling.engine
                ├── 2.7.5-SNAPSHOT
                │   ├── maven-metadata.xml
                │   ├── maven-metadata.xml.md5
                │   ├── maven-metadata.xml.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.jar
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.jar.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.jar.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.jar.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-javadoc.jar
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-javadoc.jar.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-javadoc.jar.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-javadoc.jar.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.pom
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.pom.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.pom.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1.pom.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.sha512
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.sha512.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.sha512.md5
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.sha512.sha1
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-sources.jar
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-sources.jar.asc
                │   ├── org.apache.sling.engine-2.7.5-20210505.105149-1-sources.jar.md5
                │   └── org.apache.sling.engine-2.7.5-20210505.105149-1-sources.jar.sha1
                ├── maven-metadata.xml
                ├── maven-metadata.xml.md5
                └── maven-metadata.xml.sha1
```

The signature is correct

```
$ sha512sum  org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip
02bf3f49af6c09cf7fa1defcec9665c224d665ef582e68e0aca9da5ef7cea2e63aaa8c85235b1295b61dc4ed494ca5d35ce0b0ff1ef828a07cf142565727ca7a  org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip
$ cat org.apache.sling.engine-2.7.5-20210505.105149-1-source-release.zip.sha512
02bf3f49af6c09cf7fa1defcec9665c224d665ef582e68e0aca9da5ef7cea2e63aaa8c85235b1295b61dc4ed494ca5d35ce0b0ff1ef828a07cf142565727ca7a⏎
```

**Pull request merged**